lib/kubernetes-deploy/kubernetes_resource/deployment.rb

```diff
@@ -4,78 +4,84 @@ class Deployment < KubernetesResource
+          .slice("replicas", "updatedReplicas", "availableReplicas", "unavailableReplicas"))
+        @status = @rollout_data.map { |state_replicas, num| "#{num} #{state_replicas.chop.pluralize(num)}" }.join(", ")
+      else # reset
+        @latest_rs = nil
+        @rollout_data = { "replicas" => 0 }
+        @status = nil
       end
     end
 
-    def fetch_logs
-      @representative_pod ? @representative_pod.fetch_logs : {}
-    end
-
     def fetch_events
       own_events = super
-      return own_events unless @representative_pod
-      own_events.merge(@representative_pod.fetch_events)
+      return own_events unless @latest_rs.present?
+      own_events.merge(@latest_rs.fetch_events)
+    end
+
+    def fetch_logs
+      return {} unless @latest_rs.present?
+      @latest_rs.fetch_logs
     end
 
     def deploy_succeeded?
-      return false unless @rollout_data.key?("availableReplicas")
-      # TODO: this should look at the current replica set's pods too
+      return false unless @latest_rs
+
+      @latest_rs.deploy_succeeded? &&
+      @latest_rs.desired_replicas == desired_replicas && # latest RS fully scaled up
       @rollout_data["updatedReplicas"].to_i == @rollout_data["replicas"].to_i &&
       @rollout_data["updatedReplicas"].to_i == @rollout_data["availableReplicas"].to_i
     end
 
     def deploy_failed?
-      # TODO: this should look at the current replica set's pods only or it'll never be true for rolling updates
-      @pods.present? && @pods.all?(&:deploy_failed?)
+      @latest_rs && @latest_rs.deploy_failed?
     end
 
     def deploy_timed_out?
-      # TODO: this should look at the current replica set's pods only or it'll never be true for rolling updates
-      super || @pods.present? && @pods.all?(&:deploy_timed_out?)
+      super || @latest_rs && @latest_rs.deploy_timed_out?
     end
 
     def exists?
       @found
     end
 
+    def desired_replicas
+      @definition["spec"]["replicas"].to_i
+    end
+
     private
 
-    def pod_probably_new?(pod_json)
-      return false unless @deploy_started
-      # Shitty, brittle workaround to identify a pod from the new ReplicaSet before implementing ReplicaSet awareness
-      Time.parse(pod_json["metadata"]["creationTimestamp"]) >= (@deploy_started - 30.seconds)
+    def find_latest_rs(deployment_data)
+      label_string = deployment_data["spec"]["selector"]["matchLabels"].map { |k, v| "#{k}=#{v}" }.join(",")
+      raw_json, _err, st = kubectl.run("get", "replicasets", "--output=json", "--selector=#{label_string}")
+      return unless st.success?
+
+      all_rs_data = JSON.parse(raw_json)["items"]
+      current_revision = deployment_data["metadata"]["annotations"]["deployment.kubernetes.io/revision"]
+
+      latest_rs_data = all_rs_data.find do |rs|
+        rs["metadata"]["ownerReferences"].any? { |ref| ref["uid"] == deployment_data["metadata"]["uid"] } &&
+        rs["metadata"]["annotations"]["deployment.kubernetes.io/revision"] == current_revision
```

Comment:
I'm using `ownerReferences` to be completely certain that the RS belongs to the deployment, and `revision` to select the new one, as opposed to the `n` other RS that could exists for this deployment. This seemed like the best available option based one the k8s source and issues I read, but I could be missing something. Are you aware of a better way?